pkg/drawio/layout.go

```diff
@@ -593,6 +594,106 @@ func (ly *layoutS) setIconsLocationsOnTop(square SquareTreeNodeInterface) {
+	sort.Slice(tgws, func(i, j int) bool {
+		return len(tgwOptionalCols[tgws[i]]) < len(tgwOptionalCols[tgws[j]])
+	})
+	for _, tgw := range tgws {
+		// the tgwOptionalCols are already sorted, the first and last are our last choice, so moving the first to the end:
+		tgwOptionalCols[tgw] = append(tgwOptionalCols[tgw][1:], tgwOptionalCols[tgw][0])
+		for _, ci := range tgwOptionalCols[tgw] {
+			if availableCols[ci] {
+				tgw.setLocation(newCellLocation(cloud.Location().firstRow, ly.matrix.cols[ci]))
+				delete(availableCols, ci)
+				break
+			}
+		}
+	}
+	for _, tgw := range tgws {
+		if tgw.Location() != nil {
+			continue
+		}
+		// hope we do not get here, taking the closest available:
+		var bestColAvailable int
+		bestDistance := cloud.Location().lastCol.index
+		tgwOptCol := tgwOptionalCols[tgw][0] + tgwOptionalCols[tgw][len(tgwOptionalCols[tgw])-1]/2
+		for col := range availableCols {
+			if abs(col-tgwOptCol) < bestDistance {
+				bestColAvailable = col
+				bestDistance = abs(col - tgwOptCol)
+			}
+		}
+		tgw.setLocation(newCellLocation(cloud.Location().firstRow, ly.matrix.cols[bestColAvailable]))
+	}
+}
+
+// calcTgwOptionalCols() calc the optional cols of the tgws, and a se of all the cols
+func (ly *layoutS) calcTgwOptionalCols(cloud SquareTreeNodeInterface) (
+	tgwOptionalCols map[IconTreeNodeInterface][]int, allCols map[int]bool) {
+	tgws := cloud.IconTreeNodes()
+	firstColIndex, lastColIndex := cloud.Location().firstCol.index, cloud.Location().lastCol.index
+	tgwMinCol := map[IconTreeNodeInterface]int{}
+	tgwMaxCol := map[IconTreeNodeInterface]int{}
+	tgwOptionalCols = map[IconTreeNodeInterface][]int{}
+	for _, tgw := range tgws {
+		tgwMinCol[tgw] = firstColIndex
+		tgwMaxCol[tgw] = lastColIndex
+		tgwOptionalCols[tgw] = []int{}
+	}
+	// each tgw has a MinCol and a maxCol. (the optional cols of the tgw are in this range).
+	// we iterate over the lines, and update these values:
+	for _, line := range getAllLines(ly.network) {
+		tgw := line.Router()
+		if _, ok := tgwMinCol[tgw]; ok {
+			srcLocation := line.Src().Parent().Location()
+			dstLocation := line.Dst().Parent().Location()
+			tgwMinCol[tgw] = max(tgwMinCol[tgw], min(srcLocation.firstCol.index, dstLocation.firstCol.index))
+			tgwMaxCol[tgw] = min(tgwMaxCol[tgw], max(srcLocation.lastCol.index, dstLocation.lastCol.index))
+		}
+	}
+	for _, tgw := range tgws {
+		// in the case there is no range, we flip MinCol and maxCol:
+		if tgwMinCol[tgw] > tgwMaxCol[tgw] {
+			tgwMinCol[tgw], tgwMaxCol[tgw] = tgwMaxCol[tgw], tgwMinCol[tgw]
+		}
+	}
+	// we collect the optional cols for each tgw:
+	allCols = map[int]bool{}
+	for ci := firstColIndex; ci <= lastColIndex; ci++ {
+		col := ly.matrix.cols[ci]
+		if col.width() >= iconSpace {
+			allCols[ci] = true
+			for _, tgw := range tgws {
+				if ci >= tgwMinCol[tgw] && ci <= tgwMaxCol[tgw] {
+					tgwOptionalCols[tgw] = append(tgwOptionalCols[tgw], ci)
+				}
+			}
+		}
+	}
+	return tgwOptionalCols, allCols
```

Comment:
Can we be sure here that `len(tgwOptionalCols[tgw]) > 0` for every `tgw`?

Reply:
I think can be sure, the minCol and maxCol are cols a src/dst, so we their width put them in tgwOptionalCols.
Anyway, made a fix to be on the safe side